components/org.wso2.carbon.identity.conditional.auth.functions.elk/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/elk/HttpClientManager.java

```diff
@@ -0,0 +1,227 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.conn.ssl.SSLContexts;
+import org.apache.http.conn.ssl.X509HostnameVerifier;
+import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
+import org.apache.http.impl.nio.client.HttpAsyncClients;
+import org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager;
+import org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor;
+import org.apache.http.nio.reactor.ConnectingIOReactor;
+import org.apache.http.nio.reactor.IOReactorException;
+import org.wso2.carbon.identity.application.authentication.framework.exception.FrameworkException;
+import org.wso2.carbon.identity.conditional.auth.functions.common.utils.CommonUtils;
+import org.wso2.carbon.identity.conditional.auth.functions.common.utils.Constants;
+import org.wso2.carbon.identity.conditional.auth.functions.elk.internal.ElasticFunctionsServiceHolder;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.event.IdentityEventException;
+
+import javax.net.ssl.SSLContext;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class to retrieve the HTTP Clients.
+ */
+public class HttpClientManager {
+
+    private static final Log LOG = LogFactory.getLog(HttpClientManager.class);
+
+    private static HttpClientManager instance = new HttpClientManager();
+
+    private static Map<Integer, CloseableHttpAsyncClient> clientMap = new HashMap<>();
+
+    public static HttpClientManager getInstance() {
+
+        return instance;
+    }
+
+    private HttpClientManager() {
+
+    }
+
+    /**
+     * Get HTTPClient properly configured with tenant configurations.
+     *
+     * @param tenantDomain tenant domain of the service provider.
+     * @return HttpClient
+     */
+    public CloseableHttpAsyncClient getClient(String tenantDomain) throws FrameworkException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        CloseableHttpAsyncClient client = clientMap.get(tenantId);
+
+        if (client == null) {
+
+            PoolingNHttpClientConnectionManager poolingHttpClientConnectionManager = createPoolingConnectionManager();
+
+            RequestConfig config = createRequestConfig(tenantDomain);
+
+            HttpAsyncClientBuilder httpClientBuilder = HttpAsyncClients.custom().setDefaultRequestConfig(config);
+
+            addSslContext(httpClientBuilder, tenantDomain);
+            httpClientBuilder.setConnectionManager(poolingHttpClientConnectionManager);
+
+            client = httpClientBuilder.build();
+            client.start();
+            clientMap.put(tenantId, client);
+        }
+
```

Comment:
Shall we remove the extra newlines? We only need new lines just to separate an important logic. Otherwise they will increase the length of the code

Reply:
Fixed with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/111